libafl_qemu/src/drcov.rs

```diff
@@ -0,0 +1,145 @@
+    pub fn must_instrument(&self, addr: u64) -> bool {
+        self.filter.allowed(addr)
+    }
+}
+
+impl<S> QemuHelper<S> for QemuDrCovHelper
+where
+    S: UsesInput + HasMetadata,
+{
+    fn init_hooks<QT>(&self, hooks: &QemuHooks<'_, QT, S>)
+    where
+        QT: QemuHelperTuple<S>,
+    {
+        hooks.blocks_raw(Some(gen_unique_block_ids::<QT, S>), Some(trace_block));
+    }
+
+    fn pre_exec(&mut self, _emulator: &Emulator, _input: &S::Input) {}
+
+    fn post_exec(&mut self, emulator: &Emulator, _input: &S::Input) {
+        if DRCOV_IDS.lock().unwrap().len() > self.drcov_len {
+            println!("New DrCov lenght = {}", DRCOV_IDS.lock().unwrap().len());
+            let mut drcov_vec = Vec::<DrCovBasicBlock>::new();
+            for id in DRCOV_IDS.lock().unwrap().iter() {
+                for (pc, idm) in DRCOV_MAP.lock().unwrap().iter() {
+                    if *idm == *id {
+                        let block = pc2basicblock(*pc, emulator);
+                        let mut block_len = 0;
+                        for instr in &block {
+                            block_len += instr.insn_len;
+                        }
+                        block_len -= &block.last().unwrap().insn_len;
+                        drcov_vec
+                            .push(DrCovBasicBlock::new(*pc as usize, *pc as usize + block_len));
+                    }
+                }
+            }
+
+            let mut rangemap = RangeMap::<usize, (u16, String)>::new();
+            rangemap.insert(
+                (GuestAddr::MIN as usize)..(GuestAddr::MAX as usize),
+                (0, "test".to_string()),
+            );
+            DrCovWriter::new(&rangemap)
+                .write(&self.filename, &drcov_vec)
+                .expect("Failed to write coverage file");
+        }
+        self.drcov_len = DRCOV_IDS.lock().unwrap().len();
+    }
+}
+
+pub fn gen_unique_block_ids<QT, S>(
+    hooks: &mut QemuHooks<'_, QT, S>,
+    state: Option<&mut S>,
+    pc: GuestAddr,
+) -> Option<u64>
+where
+    S: HasMetadata,
+    S: UsesInput,
+    QT: QemuHelperTuple<S>,
+{
+    if let Some(h) = hooks.helpers().match_first_type::<QemuDrCovHelper>() {
+        if !h.must_instrument(pc.into()) {
+            return None;
+        }
+    }
+    let state = state.expect("The gen_unique_block_ids hook works only for in-process fuzzing");
+    if state.metadata().get::<QemuDrCovMetadata>().is_none() {
+        state.add_metadata(QemuDrCovMetadata::new());
+    }
+    let meta = state.metadata_mut().get_mut::<QemuDrCovMetadata>().unwrap();
+
+    match DRCOV_MAP.lock().unwrap().entry(pc) {
+        Entry::Occupied(e) => {
+            let id = *e.get();
+            Some(id)
+        }
+        Entry::Vacant(e) => {
+            let id = meta.current_id;
+            e.insert(id);
+            meta.current_id = id + 1;
+            // GuestAddress is u32 for 32 bit guests
+            #[allow(clippy::unnecessary_cast)]
+            Some(id as u64)
```

Comment:
You can add here the id to drcov_ids. the translation hook is executed always before the first execution of the block. return always None so that the execution hook is never emitted, for maximum perf.

Reply:
But does a translation *always* mean that it will also be executed? I was not so sure about that as it sometimes seems to generate and not execute right after. Thats why I based my DrCov on exec and not gen

Reply:
The loop should only translate blocks that are immediately executed, afaik

Reply:
Ahhh true. I just tested that and both of you are right. The generation hooks really only happen before an execution hook. So no need to emit any execution hooks

Reply:
Then you won't get the full trace though, anyway 😅